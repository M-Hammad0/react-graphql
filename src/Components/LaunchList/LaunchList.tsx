import React from "react";
import { useQuery, gql } from "@apollo/client";
import Loading from "../Loading/Loading";
import { Waypoint } from "react-waypoint";

interface LaunchListI {
  id: string,
  launch_date_local: string;
  launch_success: boolean;
  mission_name: string;
}

interface LaunchListData {
  launches: LaunchListI[];
}

interface LaunchListVars {
  offset: number;
  limit: number;
}

export const GET_LAUNCH_INFO = gql`
  query Launches($limit: Int!, $offset: Int!) {
    launches(limit: $limit, offset: $offset) {
      id
      launch_date_local
      launch_success
      mission_name
    }
  }
`;

function LaunchList() {
  const { loading,data, fetchMore, networkStatus} = useQuery<LaunchListData, LaunchListVars>(
    GET_LAUNCH_INFO,
    {
      variables: {
        offset: 0,
        limit: 10,
      },
      notifyOnNetworkStatusChange: true,
    },
  );
  if (!data?.launches && loading) return <Loading />
  return (
    <div>
      <h3>Launches</h3>
      <div>
        {data?.launches.map((launched, i) => (
            <div key={i}>
              <p
                className="box"
                style={{ color: launched.launch_success ? "green" : "red" }}
              >
                {launched.mission_name}
              </p>
              {i === data.launches.length - 5 && (
                <Waypoint onEnter={() =>
                  fetchMore({
                    variables: {
                      offset: data?.launches.length,
                    },
                    updateQuery: (prev: any, {fetchMoreResult} ) => {
                      if(!fetchMoreResult) return prev;
                      return {
                        ...prev,
                        launches: [...prev.launches,...fetchMoreResult.launches]
                      }
                    },
                  })

                } />
              )}
            </div>
          ))}
      </div>
      {networkStatus === 3 && <h1>loading...</h1>}
    </div>
  );
}

export default LaunchList;
