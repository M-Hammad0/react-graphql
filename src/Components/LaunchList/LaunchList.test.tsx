import React from "react";
import { render, cleanup } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import LaunchList, { GET_LAUNCH_INFO } from "./LaunchList";
import Loading from "../Loading/Loading";

const mocks = [
  {
    request: {
      query: GET_LAUNCH_INFO,
      variables: {
        offset: 0,
        limit: 10,
      },
    },
    result: {
      data: {
        launches: [
          {
            id: "13",
            launch_date_local: "2014-01-06T14:06:00-04:00",
            launch_success: true,
            mission_name: "Thaicom 6",
          },
        ],
      },
    },
  },
];

afterEach(cleanup);

it("should render component without error", () => {
  const { asFragment } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <LaunchList />
    </MockedProvider>
  );
  expect(asFragment).toMatchSnapshot();
});

it("should render launch list", async () => {
  const { findByText } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <LaunchList />
    </MockedProvider>
  );
  const foo = await findByText("Thaicom 6");
  expect(foo).toBeInTheDocument();
});
