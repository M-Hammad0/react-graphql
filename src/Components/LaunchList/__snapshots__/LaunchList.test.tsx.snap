// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component without error 1`] = `[Function]`;

exports[`should render loading state initially 1`] = `[Function]`;
